packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/textFeatures.ts

```diff
@@ -0,0 +1,146 @@
+import { createRange, Position } from 'roosterjs-editor-dom';
+import { setIndentation } from 'roosterjs-editor-api';
+import {
+    BuildInEditFeature,
+    IEditor,
+    Indentation,
+    TextFeatureSettings,
+    Keys,
+    PluginKeyboardEvent,
+    SelectionRangeTypes,
+    ContentPosition,
+    PositionType,
+    ExperimentalFeatures,
+    NodePosition,
+} from 'roosterjs-editor-types';
+
+/**
+ * Requires @see ExperimentalFeatures.TabKeyTextFeatures to be enabled
+ * Provides additional functionality when press Tab:
+ *      If Whole Paragraph selected, indent paragraph,
+ *      If range is collapsed, add tab spaces
+ *      If range is not collapsed but not all the paragraph is selected, replace selection with Tab spaces
+ */
+const IndentWhenTabText: BuildInEditFeature<PluginKeyboardEvent> = {
+    keys: [Keys.TAB],
+    shouldHandleEvent: (event, editor) =>
+        editor.isFeatureEnabled(ExperimentalFeatures.TabKeyTextFeatures) &&
+        !event.rawEvent.shiftKey &&
+        !editor.getElementAtCursor('LI,TABLE', null /*startFrom*/, event),
+    handleEvent: (event, editor) => {
+        const selection = editor.getSelectionRangeEx();
+        if (selection.type == SelectionRangeTypes.Normal) {
+            editor.addUndoSnapshot(() => {
+                if (selection.areAllCollapsed) {
+                    insertTab(editor, event);
+                } else {
+                    const { ranges } = selection;
+                    const range = ranges[0];
+                    if (isWholeParagraphSelected(editor, range)) {
+                        setIndentation(editor, Indentation.Increase);
+                    } else {
+                        const tempRange = createRange(range.startContainer, range.startOffset);
+                        ranges.forEach(range => range.deleteContents());
+                        editor.select(tempRange);
+                        insertTab(editor, event);
+                    }
+                }
+            });
+
+            event.rawEvent.preventDefault();
+        }
+    },
+};
+
+/**
+ * @internal
+ */
+export const TextFeatures: Record<
+    keyof TextFeatureSettings,
+    BuildInEditFeature<PluginKeyboardEvent>
+> = {
+    indentWhenTabText: IndentWhenTabText,
+};
+
+function isWholeParagraphSelected(editor: IEditor, range: Range): boolean {
```

Comment:
What if I select two whole blocks?

Reply:
I tried in Word, if I select 1 and a half blocks, it will also indent both blocks. Can we do the same?

Reply:
I see, if we select 2 whole blocks with current implementation we can indent, but 1 and a half block we just remove the selection, I will check how to implement this.

Reply:
I just messaged Jorge to see his point of view, this is kind of weird because in Office Desktop Apps, if we select 1 and a half paragraphs we just remove all the selection (what we have in this branch), but in Office Web we do what you mentioned, we just indent.

I think that even in Word Web we just need to select 2 different blocks and on Tab it is going to indent both blocks

Reply:
Added functionality suggested, if we select more than one block and tab we indent all the blocks in selection
![Tab behavior in Listx](https://user-images.githubusercontent.com/8291124/157984277-9ef2b99e-116c-4a9a-905f-7c7f41079318.gif)
,